File: js/src/Blockchain/Blockchain.Abstraction/Workflows/EventListenerWorkflow.ts
import {
    proxyActivities,
    defineQuery,
    setHandler,
    sleep,
    continueAsNew,
    getExternalWorkflowHandle,
} from '@temporalio/workflow';
import { IBlockchainActivities } from '../Interfaces/IBlockchainActivities';
import { IUtilityActivities } from '../Interfaces/IUtilityActivities';
import { BlockRangeModel } from '../Models/BlockRangeModel';
import { lockCommitedSignal } from '../Interfaces/ISwapWorkflow';
import { IWorkflowActivities } from '../Interfaces/IWorkflowActivities';
import { TimeSpan } from '../Infrastructure/RedisHelper/TimeSpanConverter';
import { CoreTaskQueue } from '../Constants';

const blockchainActivities = proxyActivities<IBlockchainActivities>({
    startToCloseTimeout: '20s',
    scheduleToCloseTimeout: '20m',
});

const utilityActivities = proxyActivities<IUtilityActivities>({
    startToCloseTimeout: '20s',
    scheduleToCloseTimeout: '20m',
});

const workflowActivities = proxyActivities<IWorkflowActivities>({
    startToCloseTimeout: '20s',
    scheduleToCloseTimeout: '20m',
    taskQueue: CoreTaskQueue
});

let lastProcessedBlockNumber: number | undefined = undefined;
const processedTransactionHashes = new Set<string>();

export const getLastScannedBlock = defineQuery<number | undefined>('getLastScannedBlock');
export const getProcessedTransactionHashes = defineQuery<string[]>('processedTransactionHashes');

export async function EventListenerWorkflow(
    networkName: string,
    networkType: string,
    blockBatchSize: number,
    waitIntervalInSeconds: number,
    initialLastProcessedBlock?: number
): Promise<void> {
    lastProcessedBlockNumber = initialLastProcessedBlock;
    let iteration = 0;

    setHandler(getLastScannedBlock, () => lastProcessedBlockNumber);
    setHandler(getProcessedTransactionHashes, () => Array.from(processedTransactionHashes));

    while (true) {
        if (iteration >= 200) {
            await continueAsNew<typeof EventListenerWorkflow>(
                networkName,
                networkType,
                blockBatchSize,
                waitIntervalInSeconds,
                lastProcessedBlockNumber
            );
        }

        const blockData = await blockchainActivities.GetLastConfirmedBlockNumber({ NetworkName: networkName });

        if (lastProcessedBlockNumber == null) {
            lastProcessedBlockNumber = blockData.BlockNumber - blockBatchSize;
        }

        if (lastProcessedBlockNumber >= blockData.BlockNumber) {
            await sleep(TimeSpan.FromSeconds(waitIntervalInSeconds));
            iteration++;
            continue;
        }

        const blockRanges = await utilityActivities.GenerateBlockRanges(
            lastProcessedBlockNumber - 15,
            blockData.BlockNumber,
            blockBatchSize
        );

        for (const chunk of chunkArray(blockRanges, 4)) {
            await Promise.all(chunk.map(blockRange =>
                processBlockRange(networkName, blockRange)
            ));

            lastProcessedBlockNumber = chunk[chunk.length - 1].To;
        }

        iteration++;
    }
}

async function processBlockRange(
    networkName: string,
    blockRange: BlockRangeModel
): Promise<void> {
    const result = await blockchainActivities.GetEvents({
        NetworkName: networkName,
        FromBlock: blockRange.From,
        ToBlock: blockRange.To,
    });

    for (const commit of result.HTLCCommitEventMessages) {
        if (!processedTransactionHashes.has(commit.TxId)) {
            processedTransactionHashes.add(commit.TxId);
            await workflowActivities.StartSwapWorkflow(commit);
        }
    }

    for (const lock of result.HTLCLockEventMessages) {
        if (!processedTransactionHashes.has(lock.TxId)) {
            processedTransactionHashes.add(lock.TxId);
            try {

                const handle = getExternalWorkflowHandle(lock.Id);
                await handle.signal(lockCommitedSignal, lock);
            }
            catch {
            }
        }
    }
}

function chunkArray<T>(arr: T[], size: number): T[][] {
    const res: T[][] = [];
    for (let i = 0; i < arr.length; i += size) {
        res.push(arr.slice(i, i + size));
    }
    return res;
}

File: csharp/src/Workflow.Swap/Workflows/EventListenerWorkflow.cs
using Temporalio.Workflows;
using Train.Solver.Workflow.Abstractions.Models;
using Train.Solver.Infrastructure.Abstractions.Models;
using Train.Solver.Workflow.Abstractions.Activities;
using Train.Solver.Workflow.Abstractions.Workflows;
using Train.Solver.Workflow.Common.Helpers;
using Train.Solver.Workflow.Common;
using static Temporalio.Workflows.Workflow;

namespace Train.Solver.Workflow.Swap.Workflows;

[Workflow]
public class EventListenerWorkflow : IEventListenerWorkflow
{
    private ulong? _lastProcessedBlockNumber;
    private const int _maxConcurrentTaskCount = 4;
    private const int _maxIterationsBeforeContinueAsNew = 200;
    private readonly HashSet<string> _processedTransacrtionHashes = [];

    [WorkflowRun]
    public async Task RunAsync(
        string networkName,
        uint blockBatchSize,
        int waitInterval,
        ulong? lastProcessedBlockNumber = null)
    {
        _lastProcessedBlockNumber = lastProcessedBlockNumber;

        var iteration = 0;

        var network = await ExecuteActivityAsync(
            (INetworkActivities x) => x.GetNetworkAsync(networkName),
            new ActivityOptions
            {
                TaskQueue = Constants.CoreTaskQueue,
                StartToCloseTimeout = TimeSpan.FromSeconds(20),
                ScheduleToCloseTimeout = TimeSpan.FromMinutes(20),
                RetryPolicy = new()
                {
                    InitialInterval = TimeSpan.FromSeconds(5),
                    BackoffCoefficient = 1f,
                }
            });

        var solverWallets = await ExecuteActivityAsync(
            (ISwapActivities x) => x.GetRouteSourceWalletsAsync(
                network.Type),
                new()
                {
                    TaskQueue = Constants.CoreTaskQueue,
                    StartToCloseTimeout = TimeSpan.FromSeconds(20),
                    ScheduleToCloseTimeout = TimeSpan.FromMinutes(20),
                    RetryPolicy = new()
                    {
                        InitialInterval = TimeSpan.FromSeconds(5),
                        BackoffCoefficient = 1f,
                    }
                });

        while (!Temporalio.Workflows.Workflow.CancellationToken.IsCancellationRequested)
        {
            // Reset workflow history if it has been running for too long
            if (iteration >= _maxIterationsBeforeContinueAsNew)
            {
                throw CreateContinueAsNewException<EventListenerWorkflow>((x) => x.RunAsync(
                    networkName,
                    blockBatchSize,
                    waitInterval,
                    _lastProcessedBlockNumber));
            }

            try
            {
                var blockNumberWithHash = await ExecuteActivityAsync(
                    (IBlockchainActivities x) => x.GetLastConfirmedBlockNumberAsync(
                        new BaseRequest { Network = network }),
                    new()
                    {
                        TaskQueue = network.Type.ToString(),
                        StartToCloseTimeout = TimeSpan.FromSeconds(20),
                        ScheduleToCloseTimeout = TimeSpan.FromMinutes(20),
                        RetryPolicy = new()
                        {
                            InitialInterval = TimeSpan.FromSeconds(5),
                            BackoffCoefficient = 1f,
                        }
                    });

                if (!_lastProcessedBlockNumber.HasValue)
                {
                    _lastProcessedBlockNumber = blockNumberWithHash.BlockNumber - blockBatchSize;
                }

                if (_lastProcessedBlockNumber >= blockNumberWithHash.BlockNumber)
                {
                    await DelayAsync(TimeSpan.FromSeconds(waitInterval));
                    iteration++;
                    continue;
                }

                var blockRanges = await ExecuteLocalActivityAsync(
                   (IUtilityActivities x) => x.GenerateBlockRanges(
                       _lastProcessedBlockNumber.Value - 15,
                       blockNumberWithHash.BlockNumber,
                       blockBatchSize),
                   new LocalActivityOptions
                   {
                       StartToCloseTimeout = TimeSpan.FromSeconds(5)
                   });


                if (blockRanges.Any())
                {
                    foreach (var blockChunk in blockRanges.Chunk(_maxConcurrentTaskCount))
                    {
                        await Task.WhenAll(blockChunk.Select(x => ProcessBlockRangeAsync(network, solverWallets, x)));

                        _lastProcessedBlockNumber = blockChunk.Last().To;
                    }

                }
            }
            catch (Exception)
            {
                throw CreateContinueAsNewException<EventListenerWorkflow>((x) => x.RunAsync(
                    networkName,
                    blockBatchSize,
                    waitInterval,
                    _lastProcessedBlockNumber));
            }

            iteration++;
        }
    }

    private async Task ProcessBlockRangeAsync(
        DetailedNetworkDto network,
        string[] solverWallets,
        BlockRangeModel blockRange)
    {
        var result = await ExecuteActivityAsync(
            (IBlockchainActivities x) => x.GetEventsAsync(new EventRequest()
            {
                Network = network,
                FromBlock = blockRange.From,
                ToBlock = blockRange.To,
                WalletAddresses = solverWallets,
            }),
            new()
            {
                TaskQueue = network.Type.ToString(),
                StartToCloseTimeout = TimeSpan.FromSeconds(20),
                ScheduleToCloseTimeout = TimeSpan.FromMinutes(20),
                RetryPolicy = new()
                {
                    InitialInterval = TimeSpan.FromSeconds(10),
                    BackoffCoefficient = 1f,
                    MaximumAttempts = 3,
                }
            });

        foreach (var commitMessage in result.HTLCCommitEventMessages)
        {
            if (!_processedTransacrtionHashes.Add(commitMessage.TxId))
            {
                continue;
            }

            await ExecuteActivityAsync(
                (IWorkflowActivities x) => x.StartSwapWorkflowAsync(commitMessage),
                TemporalHelper.DefaultActivityOptions(Constants.CoreTaskQueue));
        }

        foreach (var lockMessage in result.HTLCLockEventMessages)
        {
            if (!_processedTransacrtionHashes.Add(lockMessage.TxId))
            {
                continue;
            }

            try
            {
                await GetExternalWorkflowHandle<ISwapWorkflow>(lockMessage.CommitId)
                    .SignalAsync((x) => x.LockCommitedAsync(lockMessage));
            }
            catch
            {
            }
        }
    }

    [WorkflowQuery]
    public ulong? GetLastScannedBlock()
    {
        return _lastProcessedBlockNumber;
    }

    [WorkflowQuery]
    public HashSet<string> ProcessedTransactionHashes()
    {
        return _processedTransacrtionHashes;
    }
}
